Add additional hooks for subclasses

jamadden opened the pull request:
Fixes #22

I see nti.site doing something like this in its subclass:
```python
class _RunJobInSite(TransactionLoop):

    def setUp(self):
        db = component.getUtility(IDatabase)
        self.conn = db.open()
        self.tearDown = self.conn.close
```

That means the object can't be used from multiple threads at the same time, but it's not used in that way anyway.


Also some minor refactoring with an eye towards speed and simplicity.

We get between 5 and 10% faster, even accounting for all the overhead of raising exceptions:

| Benchmark      | master |  this branch               |
|----------------|---------------|------------------------------|
| trivial commit | 14.9 us       | 13.2 us: 1.12x faster (-11%) |
| max retries    | 16.0 us       | 15.1 us: 1.06x faster (-5%)  |

jamadden committed:
Add additional hooks for subclasses

Fixes #22

Also some minor refactoring with an eye towards speed and simplicity.
We get between 5 and 10% faster, even accounting for all the overhead
of raising exceptions:

+----------------+---------------+------------------------------+
| Benchmark      | masterresults | issue22results               |
+================+===============+==============================+
| trivial commit | 14.9 us       | 13.2 us: 1.12x faster (-11%) |
+----------------+---------------+------------------------------+
| max retries    | 16.0 us       | 15.1 us: 1.06x faster (-5%)  |
+----------------+---------------+------------------------------+

jamadden committed:
Add test.

jamadden commented:
I need to release this project in order to make the necessary changes to nti.site. But I'd *really* like to see https://github.com/zopefoundation/transaction/pull/84 merged and released before releasing this so that we can fix #17 and remove one of [the XXX comments](https://github.com/NextThought/nti.transactions/blob/43fbfddb9a988634c746e5a74c8ce441691bd8ad/src/nti/transactions/transactions.py#L584) that explicit transactions are designed to help with in the first place.

jzuech3 commented:
I'd advocate we release this project as-is to get as much alpha burn in time as possible for these changes, coming back later for the transaction changes.

jamadden commented:
That's a fair point. The aborthook thing is pretty minor, and I think I've thought of a way to handle the XXX thing in a safe-enough way even without the transaction update; it's mostly a hypothetical issue anyway.

jamadden committed:
Distinguish more cases of bad explicit transaction management.

This lets us more confidently abort in more cases.

And know when we don't need to.
